QUR-11

Ggamid opened the pull request:
[QUR-11](https://gamidproger.atlassian.net/browse/QUR-11?atlOrigin=eyJpIjoiN2Q2ZGYwNjRiMmFmNDgyNDg3MjQ0ODVjZTc3NDNhY2IiLCJwIjoiaiJ9)

сделать чтобы первым при запуске приложения запускался экран с записью прогресса и сверху этого экрана была информация о том, на какой старнице остановилось чтение

Ggamid committed:
HomeView
Сделал чтобы первым при запуске приложения отображался экран для записи прогресс

Ggamid committed:
getSurahName
Написал метод который возвращает название первой суры на странице

Ggamid committed:
Отображение текущей позиции чтения
Отображение текущей позиции чтения в виде номера последней страницы чтения и названия первой суры на этой странице

gasanovrt commented on QuranTracker/Views/HomeView.swift:
```diff
@@ -10,8 +10,8 @@ import SwiftfulUI
+    @State var writeProgressOffset: Int = 0
+    @State var blurRadius: CGFloat = 7
     @State private var isNextScreenActive = false
 
```
writeProgressOffset и blurRadius используются где то извне? иначе private

Ggamid replied:
сделал поля  writeProgressOffset, blurRadius private В HomeView

gasanovrt commented on QuranTracker/Views/WriteProgress/WriteProgress-ViewModel.swift:
```diff
@@ -7,6 +7,7 @@
 
```
логичнее назвать файл
WriteProgress+ViewModel.swift

Ggamid replied:
сделал

gasanovrt commented on QuranTracker/Views/WriteProgress/WriteProgress-ViewModel.swift:
```diff
@@ -16,17 +17,20 @@ extension WriteProgressView {
         var startPage: Int = 0
```
final class

Ggamid replied:
сделал

gasanovrt commented on QuranTracker/Views/WriteProgress/WriteProgress-ViewModel.swift:
```diff
@@ -16,17 +17,20 @@ extension WriteProgressView {
         var presentAlert = false
+        var currentPage: Int?
+        var currentSurah: String?
 
```
проверь какие методы и свойства можно сделать private

Ggamid replied:
свойство никакие нельзя, а метод checkNumbers сделал private

gasanovrt commented on QuranTracker/Views/WriteProgress/WriteProgress-ViewModel.swift:
```diff
@@ -35,6 +39,22 @@ extension WriteProgressView {
             return false
```
тут как понимаю целевое то что внутри блока checkNumbers. значит надо
```
func saveReadingSession(modelContext: ModelContext) -> Bool {
    guard checkNumbers() else {
        presentAlert = true
        return false
    }

    let readingSession = QuranReadingSession(
        startPage: self.startPage,
        endPage: self.endPage,
        sessionDate: self.sessionDate
    )
    modelContext.insert(readingSession)

    startPage = 0
    endPage = 0
    return true
}
```

Ggamid replied:
сделал

gasanovrt commented on QuranTracker/Views/WriteProgress/WriteProgress-ViewModel.swift:
```diff
@@ -35,6 +39,22 @@ extension WriteProgressView {
+        func getSurahName(quranReadingSession: QuranReadingSession) async {
+            currentPage = quranReadingSession.endPage
+            do {
+                guard let currentPage else { return }
```
мб return как то отработать?

Ggamid replied:
Имеешь ввиду добавить enum с возможными ошибками и их выкидывать?

Ggamid replied:
ничего в голову не приходит

gasanovrt commented on QuranTracker/Views/WriteProgress/WriteProgress-ViewModel.swift:
```diff
@@ -35,6 +39,22 @@ extension WriteProgressView {
+            do {
+                guard let currentPage else { return }
+                let quranPage = try await QuranData.fetchData(about: currentPage)
+                let surah = quranPage.surahs.sorted { Int($0.key)! < Int($1.key)! } // сортируем массив по номеру сур
```
никаких форс анрапов

Ggamid replied:
Исправил на 
`let surah = quranPage.surahs.sorted {
                    guard let firstKey = Int($0.key), let secondKey = Int($1.key) else { return false }
                        return firstKey < secondKey
                }`

gasanovrt commented on QuranTracker/Views/WriteProgress/WriteProgress-ViewModel.swift:
```diff
@@ -35,6 +39,22 @@ extension WriteProgressView {
+                currentSurah = surah.first?.value.englishName
+                return
+            } catch {
+                print("something went wrong")
```
вроде в catch можно еррор получить

Ggamid replied:
добавил обработку ошибок FetchingDataError

gasanovrt commented on QuranTracker/Views/WriteProgress/WriteProgressView.swift:
```diff
@@ -32,36 +32,49 @@ struct WriteProgressView: View {
+
+                                    DispatchQueue.main.asyncAfter(deadline: .now() + 1.5) { onXmarkPressed?() }
+                                    amountIsFouced = false
                                 }
```
экшен вынеси в метод

Ggamid replied:
Перенес логику скрытия экрана WriteProgressView в отдельны метод closeWriteProgressViewWithDelay

gasanovrt commented on QuranTracker/Views/WriteProgress/WriteProgressView.swift:
```diff
@@ -32,36 +32,49 @@ struct WriteProgressView: View {
+            Task {
+                guard let lastReadingSession = readingSessions.last else { return }
+                await vm.getSurahName(quranReadingSession: lastReadingSession)
+            }
```
тут и в ButtonQT(text: "Сохранить")  есть такая таска. вынеси в одну

Ggamid replied:
Перенес логику получения информации о текущей позиции чтения в отдельный метод getInformationAboutCurrentPosition

gasanovrt commented on QuranTracker/Views/WriteProgress/WriteProgressView.swift:
```diff
@@ -170,4 +182,24 @@ extension WriteProgressView {
                 .frame(maxWidth: 30)
         }
     }
+    private var currentPosition: some View {
```
логичнее тут использовать вью билдер

Ggamid replied:
а какая разница?

gasanovrt replied:
в таком случае currentPosition отрисуется только если сработает условие if let currentPage = vm.currentPage, let currentSurah = vm.currentSurah {

 
```
@ViewBuilder private var currentPosition: some View {
    if let currentPage = vm.currentPage, let currentSurah = vm.currentSurah {
        VStack {
                Text("Текущая позиция")
                    .font(.title2)
                    .fontWeight(.bold)
                HStack {
                    VStack {
                        Text("страница.")
                        Text("\(currentPage)")
                    }
                    VStack {
                        Text("Сура")
                        Text(currentSurah)
                    }
                }
        }
        .frame(height: 100)
    }
}

```

Ggamid replied:
В ветке QUR-6 исправил

Ggamid committed:
HomeView, WriteProgress+ViewModel changes
Изменения в соответствии с комментариями
сделал поля  writeProgressOffset, blurRadius private В HomeView
WriteProgress-ViewModel переименовал в WriteProgress+ViewModel и сделал класс ViewModel final
изменил метод saveReadingSession
Избавился от форс анврапа в getSurahName и добавил обработку ошибок FetchingDataError
checkNumbers -> private checkNumbers

Ggamid committed:
WriteProgressView
Перенес логику получения информации о текущей позиции чтения в отдельный метод getInformationAboutCurrentPosition
Перенес логику скрытия экрана WriteProgressView в отдельны метод closeWriteProgressViewWithDelay
